node/src/components/consensus/era_supervisor.rs

```diff
@@ -1333,3 +1358,58 @@ pub(crate) fn oldest_bonded_era(protocol_config: &ProtocolConfig, current_era: E
         .saturating_sub(bonded_eras(protocol_config))
         .max(protocol_config.last_activation_point)
 }
+
+async fn check_deploys_for_replay_and_validate_block<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    proto_block_era_id: EraId,
+    sender: I,
+    proto_block: ProtoBlock,
+    timestamp: Timestamp,
+) -> Event<I>
+where
+    REv: From<BlockValidationRequest<ProtoBlock, I>> + From<StorageRequest>,
+    I: Clone + Send + 'static,
+{
+    for deploy_hash in proto_block.deploys_iter() {
+        let execution_results = match effect_builder
+            .get_deploy_and_metadata_from_storage(*deploy_hash)
+            .await
+        {
+            None => continue,
+            Some((_, DeployMetadata { execution_results })) => execution_results,
+        };
+        for (block_hash, _) in execution_results {
+            match effect_builder
+                .get_block_header_from_storage(block_hash)
+                .await
+            {
+                // TODO: Throw an error here and use fatal! elsewhere
```

Comment:
Until then, seems like a better idea will be to simply panic. Otherwise we're risking replaying a deploy after all